### PR TITLE
Fix dangling reference

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #3942

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Review by pietro68 — APPROVED** [review]
Code changes and reason for them understood.

**Comment by Paul-Licameli** [comment]
Added another commit with similar precautions for VST3 and requested another review.

File: src/effects/VST3/VST3Wrapper.cpp [hunk 1]
```diff
@@ -447,9 +448,19 @@ void VST3Wrapper::FetchSettings(EffectSettings& settings)
       auto cleanup = finally([&] { componentHandler->EndStateChange(); });
 
       //Restore state
-      const auto* vst3settings = &GetSettings(settings);
-      if(!vst3settings->processorState.has_value())
-         vst3settings = &GetSettings(mDefaultSettings);
+      // See Issue 3942
+      // Callbacks to performEdit might happen inside setState that call
+      // ModifySettings to reassign the given settings, but the settings may be
```
Comment (vsverchinsky) [inline src/effects/VST3/VST3Wrapper.cpp:453]: `BeginStateChange` prevents `ModifySettings` from being called in the same callstack with `setState`

Comment (Paul-Licameli) [reply src/effects/VST3/VST3Wrapper.cpp:453]: You are right.

Do you recall why you wrote BeginStateChange?

Were you observing exactly the sort of dangling reference problem I destribed?

**Pull request merged**